tiledb/api/src/array/dimension/mod.rs

```diff
@@ -207,62 +209,63 @@ impl<'ctx> Builder<'ctx> {
         })
     }
 
-    /// N.B., This API will eventually be removed when we fixup the current
-    /// implementation of `new` to take a `DimensionConstraint` argument instead
-    /// of the serde values for range and extent.
-    pub fn new_string(
+    pub fn new_optional(
         context: &'ctx Context,
         name: &str,
         datatype: Datatype,
+        domain: Option<SingleValueRange>,
+        extent: Option<Extent>,
     ) -> TileDBResult<Self> {
         let c_datatype = datatype.capi_enum();
+
         let c_name = cstring!(name);
-        let mut c_dimension: *mut ffi::tiledb_dimension_t =
-            std::ptr::null_mut();
 
-        context.capi_call(|ctx| unsafe {
-            ffi::tiledb_dimension_alloc(
-                ctx,
-                c_name.as_ptr(),
-                c_datatype,
-                std::ptr::null(),
-                std::ptr::null(),
-                &mut c_dimension,
-            )
-        })?;
-        Ok(Builder {
-            dim: Dimension {
-                context,
-                raw: RawDimension::Owned(c_dimension),
-            },
-        })
-    }
+        let mut domain_data: Vec<u8> = Vec::new();
```

Comment:
Ok I think I follow why you need `Vec` here - because of the enum variants not using a slice, it might not be guaranteed that the values are contiguous, this fixes things in memory the way you want.

Regarding the endian-ness... does tiledb use the native byte ordering?
If it does, then converting may be incorrect.
If it does not, then we probably have a lot more work to do to get things to line up.

For the extent, seems like you could do `value as *const DT as *const std::ffi::c_void` and avoid a vec there.

Reply:
AFAIK, TileDB just assumes the entire world is little endian. I don't think there's anything in core that tries to handle running on big endian architectures. Although I have found a few places where we store data on disk as big endian for reasons I have yet to fathom. (If memory serves, one of the filter pipeline option values is stored big endian).

For the Vec, that's pretty much correct. I needed contiguous bytes in a type safe allocation that outlives the call to `tiledb_dimension_alloc` and that was the least worse way I could think to do it. Though I'll likely revisit this after I finish figuring out that bug that proptests found yesterday.